### DISCUSSION TITLE
Unable to set customer Phone and Email in order - PHP SDK

### DISCUSSION THREAD

**Discussion by cleoni** [fattureincloud/api#34] [Q&A] [answered]
Sto lavorando sulla trasmissione di un ordine via PHP. L'esempio sulla documentazione funziona ma a parte che è basato su un documento diverso (invoice) non copre tutti i casi possibili. Dunque sto cercando di aggiungere nell'ordine il telefono e l'email del cliente, cosa che a me serve e che si faceva con la vecchia API.

Purtroppo non trovo la lista dei metodi disponibili per "Entity" per cui guardo nel sorgente della libreria, vedo che è possibile fare questo:

$entity->setPhone("051123456");
$entity->setEmail("prova@prova.it");

in questo modo dovrei essere riuscito a settare telefono e email del cliente intestatario dell'ordine che sto spedendo.
tutto il resto segue come nell'esempio:

$order = new IssuedDocument;
		$order->setType(IssuedDocumentType::ORDER);
		$order->setEntity($entity);
// Trasmissione di $order

però poi mi trovo su fattureincloud il documento "ordine" senza i campi mail e telefono, mentre la versione creata con la precedente api v1 li ha.
Noto anche che l'oggetto response ritornato dalla chiamata a $apiInstance->createIssuedDocument ha i dati del cliente come li ha creati, in teoria dovrebbero essere quelli che ho spedito, ma che contiene invece valori vuoti per email e telefono:

..
                                            [email] => 
                                            [certified_email] => 
                                            [phone] => 
                                            [fax] => 
..

quindi la domanda è: come fare se voglio scrivere nell'ordine mail e telefono del cliente? Ho fatto bene a usare setPhone e setEmail o si doveva fare in modo diverso?

A margine di questo, la nuova API (che è sicuramente molto "elegante" della precedente) nasconde però la struttura "data" che contiene tutti i dati passati poi effettivamente all'endpoint.
Sarebbe meno elegante - ma assai più pratico - poter inviare all'API la struttura "data" così come si vede nei vari esempi in modo da essere sicuri di quello che effettivamente viene spedito. Sarebbe possibile farlo?

**Comment by valmoz** [comment, marked as answer]
Buongiorno,
non ci torna quello che sta dicendo. Lei vede il numero e la mail all'interno del PDF del documento oppure tramite API V1? può farci degli screenshot, possibilmente con dati d'esempio? (per inviare dati reali è necessario contattare il supporto clienti, questa è una discussione pubblica).

Sulle V2 quei dati vengono salvati solamente quando si scrive in anagrafica. Poiché al momento non abbiamo inserito la funzionalità di aggiorna anagrafica sulla create issued document, l'API ignora quei campi.

Le suggerisco quindi di provare ad effettuare le seguenti operazioni:
- Utilizzare il metodo [Create Client](https://developers.fattureincloud.it/api-reference#post-/c/-company_id-/entities/clients) per aggiungere l'utente in anagrafica. Qui lei potrà aggiungere anche numero e email
- Nella Create Issued Document, può aggiungere l'ID alla entity, in modo tale da legare la fattura al client. 

Ci faccia sapere se in questo modo sta ottenendo il risultato desiderato.

Grazie
Mauro

**Reply by cleoni** [reply]
Salve, potrei mandarvi qualche screenshot ma intanto vi do qualche indicazione.

Quello che ho fatto è inviare un ordine con dati del cliente "al volo". Non sto creando il cliente in anagrafica e il perché è presto detto: sto facendo un upgrade da APIv1 a APIv2 e sto cercando di cambiare meno cose possibile per contenere i costi.

Con APIv1 la creazione di ordine con dati cliente estemporanei funzionava (e infatti ho questo ordine creato con APIv1 dove i dati ci sono) mentre con APIv2 qualcosa non sembra andare per il verso giusto, io tramite API cerco di settare telefono ed email ma poi quello che viene trasmesso al backend non contiene questi dati.

Ora verifico se possiamo sentirci telefonicamente.

Eventualmente posso postare il codice che uso, il problema è abbastanza facile da replicare.

**Reply by valmoz** [reply]
Ciao @cleoni 
questo è un comportamento atteso dell'API v2.

Poiché che nel documento generato non appaiono quei dati per il cliente (è questo che non ci torna, lei dice che invece con le V1 appaiono, se non ho capito male), abbiamo deciso di ignorarli totalmente per le entity che non vengono registrate in anagrafica ma solamente inserite al momento della creazione della fattura. Per questo le suggerivo di effettuare prima la chiamata Create Client.

Nell'SDK lei continua a vedere questi campi perché nella generazione dell'SDK abbiamo usato lo stesso modello per le chiamate di aggiornamento anagrafica e per dichiarare una entity nella generazione del documento, perché di fatto i due modelli differiscono solo di questi campi extra.

**Reply by cleoni** [reply]
Va bene, almeno ora è chiaro che "non è un bug, è una feature". Per fortuna ho appena verificato con i titolari, questi due dati non sono essenziali in fattura.

Questa richiesta la possiamo ritenere chiusa.
